File: Advanced/PageAndRDLX/CustomPdfExport/C#/PdfRendering/Rendering/PdfGenerator.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using GrapeCity.ActiveReports.Extensibility.Rendering;
using GrapeCity.ActiveReports.Extensibility.Rendering.Components;
using GrapeCity.ActiveReports.Drawing;
using GrapeCity.ActiveReports.Rendering;
using GrapeCity.ActiveReports.Rendering.Export;
using PdfSharp.Drawing;
using PdfSharp.Pdf;
using GrapeCity.ActiveReports.Drawing.Core;

namespace ActiveReports.Samples.Export.Rendering.PdfSharp
{
	/// <summary>
	/// PDF generator based on PDFsharp: http://www.pdfsharp.net/
	/// </summary>
	internal sealed class PdfGenerator : IGenerator, IDisposable
	{
		private static readonly ITextMetricsProvider GcMetricsProvider = new TextMetricsProvider(FontsFactory.Instance, new TextMetricsProvider.TextSettings());
		private static readonly IRenderersFactory GcRenderersFactory = new DocRenderersFactory();

		private readonly Stream _outputStream;

		private PdfDocument _document;
		private readonly IList<XGraphics> _graphics = new List<XGraphics>();
		private readonly PdfImagesFactory _images;
		private readonly PdfFontsFactory _fonts;
		private PdfInteractivityRenderer _interactivityRenderer;

		public PdfGenerator(Stream outputStream, bool embedFonts)
		{
			_outputStream = outputStream;
			_images = new PdfImagesFactory();
			_fonts = new PdfFontsFactory(embedFonts);
		}

		void IDisposable.Dispose()
		{
			foreach (var g in _graphics)
			{
				Debug.Assert(g.GraphicsStateLevel == 0);
				g.Dispose();
			}
			_graphics.Clear();
			if (_document != null)
			{
				_document.Save(_outputStream);
				_document.Dispose();
			}
			_document = null;
			((IDisposable)_images).Dispose();
		}

		public ITextMetricsProvider MetricsProvider => GcMetricsProvider;

		public IRenderersFactory RenderersFactory => GcRenderersFactory;

		public IInteractivityRenderer InteractivityRenderer => _interactivityRenderer;

		void IGenerator.Init(IReport report)
		{
			_document = new PdfDocument { Version = 17 };
			_interactivityRenderer = new PdfInteractivityRenderer(_document, _graphics);
		}

		IDrawingCanvas IGenerator.NewPage(int pageNumber, SizeF pageSize)
		{
			var page = new PdfPage(_document);
			page.Width = new XUnit(pageSize.Width / PdfConverter.TwipsPerPoint, XGraphicsUnit.Point);
			page.Height = new XUnit(pageSize.Height / PdfConverter.TwipsPerPoint, XGraphicsUnit.Point);
			_document.AddPage(page);
			var graphics = XGraphics.FromPdfPage(page);
			_graphics.Add(graphics);
			return new PdfContentGenerator(graphics, _images, _fonts);
		}
	}
}

File: Advanced/PageAndRDLX/CustomPdfExport/VB.NET/PdfRendering/Rendering/PdfGenerator.vb
Imports PdfSharp.Pdf
Imports PdfSharp.Drawing
Imports GrapeCity.ActiveReports.Rendering.Export
Imports GrapeCity.ActiveReports.Rendering
Imports GrapeCity.ActiveReports.Drawing
Imports GrapeCity.ActiveReports.Extensibility.Rendering.Components
Imports GrapeCity.ActiveReports.Extensibility.Rendering
Imports System.IO
Imports System.Drawing
Imports GrapeCity.ActiveReports.Drawing.Core

''' <summary>
''' PDF generator based on PDFsharp: http://www.pdfsharp.net/
''' </summary>
Friend NotInheritable Class PdfGenerator
	Implements IGenerator
	Implements IDisposable
	Private Shared ReadOnly GcMetricsProvider As ITextMetricsProvider = New TextMetricsProvider(FontsFactory.Instance, New TextMetricsProvider.TextSettings())
	Private Shared ReadOnly GcRenderersFactory As IRenderersFactory = New DocRenderersFactory()

	Private ReadOnly _outputStream As Stream

	Private _document As PdfDocument
	Private ReadOnly _graphics As IList(Of XGraphics) = New List(Of XGraphics)()
	Private ReadOnly _images As PdfImagesFactory
	Private ReadOnly _fonts As PdfFontsFactory
	Private _interactivityRenderer As PdfInteractivityRenderer

	Public Sub New(outputStream As Stream, embedFonts As Boolean)
		_outputStream = outputStream

		_images = New PdfImagesFactory()
		_fonts = New PdfFontsFactory(embedFonts)
	End Sub

	Sub Dispose() Implements IDisposable.Dispose
		For Each g As XGraphics In _graphics
			Debug.Assert(g.GraphicsStateLevel = 0)
			g.Dispose()
		Next
		_graphics.Clear()
		If _document IsNot Nothing Then
			_document.Save(_outputStream)
			_document.Dispose()
		End If
		_document = Nothing
		CType(_images, IDisposable).Dispose()
	End Sub

	Public ReadOnly Property MetricsProvider() As ITextMetricsProvider Implements IGenerator.MetricsProvider

		Get
			Return GcMetricsProvider
		End Get
	End Property

	Public ReadOnly Property RenderersFactory() As IRenderersFactory Implements IGenerator.RenderersFactory
		Get
			Return GcRenderersFactory
		End Get
	End Property
	
	Public ReadOnly Property InteractivityRenderer As IInteractivityRenderer Implements IGenerator.InteractivityRenderer
		Get
			Return _interactivityRenderer
		End Get
	End Property

	Sub Init(report As IReport) Implements IGenerator.Init
		_document = New PdfDocument() With {
				.Version = 17
			}
		_interactivityRenderer = New PdfInteractivityRenderer(_document, _graphics)
	End Sub

	Function NewPage(pageNumber As Integer, pageSize As SizeF) As IDrawingCanvas Implements IGenerator.NewPage
		Dim page As New PdfPage(_document)
		page.Width = New XUnit(pageSize.Width / PdfConverter.TwipsPerPoint, XGraphicsUnit.Point)
		page.Height = New XUnit(pageSize.Height / PdfConverter.TwipsPerPoint, XGraphicsUnit.Point)
		_document.AddPage(page)
		Dim graphics = XGraphics.FromPdfPage(page)
		_graphics.Add(graphics)
		Return New PdfContentGenerator(graphics, _images, _fonts)
	End Function
	
End Class
